src/core/meshcop/border_agent.cpp

```diff
@@ -252,11 +312,116 @@ BorderAgent::BorderAgent(Instance &aInstance)
+{
+    Message *        message = NULL;
+    Ip6::MessageInfo messageInfo;
+    uint16_t         offset;
+    otError          error;
+
+    OT_UNUSED_VARIABLE(aHeader);
+
+    {
+        UdpEncapsulationTlv tlv;
+
+        SuccessOrExit(error = Tlv::GetOffset(aMessage, Tlv::kUdpEncapsulation, offset));
+        aMessage.Read(offset, sizeof(tlv), &tlv);
+
+        VerifyOrExit((message = GetInstance().GetIp6().GetUdp().NewMessage(0)) != NULL, error = OT_ERROR_NO_BUFS);
+        SuccessOrExit(error = message->SetLength(tlv.GetUdpLength()));
+        aMessage.CopyTo(offset + sizeof(tlv), 0, tlv.GetUdpLength(), *message);
+
+        messageInfo.SetSockPort(tlv.GetSourcePort() != 0 ? tlv.GetSourcePort()
+                                                         : GetInstance().GetIp6().GetUdp().GetEphemeralPort());
+        messageInfo.SetSockAddr(mCommissionerAloc.GetAddress());
+        messageInfo.SetPeerPort(tlv.GetDestinationPort());
+    }
+
+    {
+        IPv6AddressTlv tlv;
+
+        SuccessOrExit(error = Tlv::Get(aMessage, Tlv::kIPv6Address, sizeof(tlv), tlv));
+        messageInfo.SetPeerAddr(tlv.GetAddress());
+    }
+
+    SuccessOrExit(error = GetInstance().GetIp6().GetUdp().SendDatagram(*message, messageInfo, Ip6::kProtoUdp));
+    otLogInfoMeshCoP(GetInstance(), "Proxy transmit sent");
+
+exit:
+    if (error != OT_ERROR_NONE)
+    {
+        otLogWarnMeshCoP(GetInstance(), "Failed to send proxy stream: %s", otThreadErrorToString(error));
+
+        if (message != NULL)
+        {
+            message->Free();
+        }
+    }
+}
+
+bool BorderAgent::HandleProxyMessage(const Message &aMessage, const Ip6::MessageInfo &aMessageInfo)
+{
+    Coap::Header header;
+    otError      error;
+    Message *    message = NULL;
+    ThreadNetif &netif   = GetNetif();
+
+    VerifyOrExit(aMessageInfo.GetSockAddr() == mCommissionerAloc.GetAddress(),
+                 error = OT_ERROR_DESTINATION_ADDRESS_FILTERED);
+
+    VerifyOrExit(aMessage.GetLength() > 0, error = OT_ERROR_NONE);
+
+    header.Init(OT_COAP_TYPE_NON_CONFIRMABLE, OT_COAP_CODE_POST);
+    header.AppendUriPathOptions(OT_URI_PATH_PROXY_RX);
+    header.SetPayloadMarker();
+
+    VerifyOrExit((message = NewMeshCoPMessage(netif.GetCoapSecure(), header)) != NULL, error = OT_ERROR_NO_BUFS);
+
+    {
+        UdpEncapsulationTlv tlv;
+        uint16_t            offset;
+        uint16_t            udpLength = aMessage.GetLength() - aMessage.GetOffset();
+
+        tlv.Init();
+        tlv.SetSourcePort(aMessageInfo.GetPeerPort());
+        tlv.SetDestinationPort(aMessageInfo.GetSockPort());
+        tlv.SetUdpLength(udpLength);
+        SuccessOrExit(error = message->Append(&tlv, sizeof(tlv)));
+
+        offset = message->GetLength();
+        SuccessOrExit(error = message->SetLength(offset + udpLength));
+        VerifyOrExit(aMessage.CopyTo(aMessage.GetOffset(), offset, udpLength, *message) == udpLength,
```

Comment:
assert here instead?  `SetLength()` should allocate the appropriate message buffer.